plan/new_physical_plan_builder.go

```diff
@@ -497,15 +497,16 @@ func (p *baseLogicalPlan) getBestTask(bestTask task, prop *requiredProp, pp Phys
 	return bestTask, nil
 }
 
-func tryToAddUnionScan(cop *copTask, conds []expression.Expression, ctx context.Context, allocator *idAllocator) task {
-	if ctx.Txn() == nil || ctx.Txn().IsReadOnly() {
-		return cop
-	}
+func addUnionScan(ctx context.Context, cop *copTask, conds []expression.Expression, needColHandle bool, allocator *idAllocator) task {
 	task := finishCopTask(cop, ctx, allocator)
 	us := PhysicalUnionScan{
-		Conditions: conds,
+		Conditions:    conds,
+		NeedColHandle: needColHandle,
 	}.init(allocator, ctx)
-	us.SetSchema(task.plan().Schema())
+	us.SetSchema(task.plan().Schema().Clone())
+	if !us.NeedColHandle {
```

Comment:
Order of schema.Columns is the same with row.Data the executor returns. So if we don't trim it, all the plan above will need to record the handle.